pexpect/expect.py

```diff
@@ -6,45 +6,95 @@ class Expecter(object):
     def __init__(self, spawn, searcher, searchwindowsize=-1):
         self.spawn = spawn
         self.searcher = searcher
+        # A value of -1 means to use the figure from spawn, which should
+        # be None or a positive number.
         if searchwindowsize == -1:
             searchwindowsize = spawn.searchwindowsize
         self.searchwindowsize = searchwindowsize
+        self.lookback = None
+        if hasattr(searcher, 'longest_string'):
+            self.lookback = searcher.longest_string
 
     def new_data(self, data):
         spawn = self.spawn
         searcher = self.searcher
 
-        pos = spawn._buffer.tell()
-        spawn._buffer.write(data)
-        spawn._before.write(data)
-
-        # determine which chunk of data to search; if a windowsize is
-        # specified, this is the *new* data + the preceding <windowsize> bytes
-        if self.searchwindowsize:
-            spawn._buffer.seek(max(0, pos - self.searchwindowsize))
-            window = spawn._buffer.read(self.searchwindowsize + len(data))
+        if data is None:
```

Comment:
It is pre-existing that this code is called from _async.py and expect.py before any data, and again with the new data.  I am refactoring to be two methods and update both callers.  Sorry if it looks like I'm not updating the code here; I'm working on this on an internal copy of the code within our company, and then will push to my github branch and update review.

Reply:
The split of new_data has been completed, additional tests (demonstrating a few of the bugs this fixes -- it fixes a lot of bugs, though, not all of them easy to test) are the outstanding action item here.

Key bugs this fixes to cover in tests, potentially:
- truncated output: this is the big data corruption bug which people keep reporting and which we ran into.
- not matching when we should, when searchwindowsize is higher in one expecter than the previous expecter on a given spawn.
- matching when we shouldn't, when a match occurs outside searchwindowsize.

Reply:
First test case of the three I proposed is now added to the pull request.

Reply:
Added second above-mentioned regression test.

Reply:
Added third/final regression test.

May also be useful to document which version introduced each of these three bugs, rather than just mentioning they are all present in 4.7.  I know none of them were present in 3.3, but that's a wide range...

Reply:
Looks like all three bugs were introduced in commit fd7332f59e003a8ee658647f883d4b03018c5e5b on Jan 15, 2018.

Reply:
Made the docstrings state these three bugs were all present in releases 4.4 - 4.7.